src/preferences/colorpaletteeditor.cpp

```diff
@@ -0,0 +1,199 @@
+#include "preferences/colorpaletteeditor.h"
+
+#include <QColorDialog>
+#include <QDialogButtonBox>
+#include <QHBoxLayout>
+#include <QHeaderView>
+#include <QLabel>
+#include <QMenu>
+#include <QStandardItemModel>
+#include <QTableView>
+
+#include "preferences/colorpalettesettings.h"
+#include "util/color/predefinedcolorpalettes.h"
+
+namespace {
+const QColor kDefaultPaletteColor(0, 0, 0);
+}
+
+ColorPaletteEditor::ColorPaletteEditor(QWidget* parent)
+        : QWidget(parent),
+          m_bPaletteExists(false),
+          m_bPaletteIsReadOnly(false),
+          m_pPaletteNameComboBox(make_parented<QComboBox>()),
+          m_pTableView(make_parented<QTableView>()),
+          m_pModel(make_parented<ColorPaletteEditorModel>()) {
+    m_pPaletteNameComboBox->setEditable(true);
+
+    QDialogButtonBox* pButtonBox = new QDialogButtonBox();
+    m_pSaveButton = pButtonBox->addButton(QDialogButtonBox::Save);
+    m_pResetButton = pButtonBox->addButton(QDialogButtonBox::Reset);
+    m_pDiscardButton = pButtonBox->addButton(QDialogButtonBox::Discard);
+
+    QHBoxLayout* pTopLayout = new QHBoxLayout();
+    pTopLayout->addWidget(new QLabel("Name:"));
+    pTopLayout->addWidget(m_pPaletteNameComboBox, 1);
+    pTopLayout->addWidget(pButtonBox);
+
+    QVBoxLayout* pLayout = new QVBoxLayout();
+    pLayout->addLayout(pTopLayout);
+    pLayout->addWidget(m_pTableView, 1);
+    setLayout(pLayout);
+    setContentsMargins(0, 0, 0, 0);
+
+    // Set up model
+    m_pModel->setColumnCount(2);
+    m_pModel->setHeaderData(0, Qt::Horizontal, tr("Color"), Qt::DisplayRole);
+    m_pModel->setHeaderData(1, Qt::Horizontal, tr("Assign to Hotcue"), Qt::DisplayRole);
+    connect(m_pModel,
+            &ColorPaletteEditorModel::dirtyChanged,
+            this,
+            &ColorPaletteEditor::slotUpdateButtons);
+    connect(m_pModel,
+            &ColorPaletteEditorModel::emptyChanged,
+            this,
+            &ColorPaletteEditor::slotUpdateButtons);
+
+    // Setup up table view
+    m_pTableView->setShowGrid(false);
+    m_pTableView->setSelectionBehavior(QAbstractItemView::SelectRows);
+    m_pTableView->setSelectionMode(QAbstractItemView::SingleSelection);
+    m_pTableView->setDragDropMode(QAbstractItemView::InternalMove);
+    m_pTableView->setDragDropOverwriteMode(false);
+    m_pTableView->setContextMenuPolicy(Qt::CustomContextMenu);
+    m_pTableView->setModel(m_pModel);
+
+    connect(m_pTableView,
+            &QTableView::doubleClicked,
+            [this](const QModelIndex& index) {
+                if (index.isValid() && index.column() == 0) {
+                    QColor color = QColorDialog::getColor();
+                    if (color.isValid()) {
+                        m_pModel->setColor(index.row(), color);
+                    }
+                }
+            });
+
+    connect(m_pTableView,
+            &QTableView::customContextMenuRequested,
+            [this](const QPoint& pos) {
```

Comment:
please only only use lambdas for very short, trivial functions.  This is long enough that it should be a separate handler.

Reply:
(There are a lot of handlers here, nearly all of them are fairly long and complicated. If they are broken into functions then we might actually be able to test them)